File: src/main/kotlin/com/huawei/cangjie1/psi/CjDeclarationWithBody.java
package com.huawei.cangjie1.psi;

import com.intellij.psi.PsiElement;
import org.jetbrains.annotations.NotNull;
import org.jetbrains.annotations.Nullable;

import java.util.List;


public interface CjDeclarationWithBody extends CjDeclaration {
    @Nullable
    CjExpression getBodyExpression();

    @Nullable
    PsiElement getEqualsToken();

    @Override
    @Nullable
    String getName();



    boolean hasBlockBody();

    boolean hasBody();

    boolean hasDeclaredReturnType();

    @NotNull
    List<CjParameter> getValueParameters();

    @Nullable
    default CjBlockExpression getBodyBlockExpression() {
        CjExpression bodyExpression = getBodyExpression();
        if (bodyExpression instanceof CjBlockExpression) {
            return (CjBlockExpression) bodyExpression;
        }

        return null;
    }
}

File: src/main/kotlin/com/linqingying/cangjie/psi/CjDeclarationWithBody.kt
package com.linqingying.cangjie.psi

import com.intellij.psi.PsiElement

interface CjDeclarationWithBody : CjDeclaration {

    val bodyExpression: CjExpression?

    val equalsToken: PsiElement?

    override fun getName(): String?


    fun hasBlockBody(): Boolean

    fun hasBody(): Boolean

    fun hasDeclaredReturnType(): Boolean


    val valueParameters: List<CjParameter >

    val bodyBlockExpression: CjBlockExpression?
        get() {
            val bodyExpression = bodyExpression
            if (bodyExpression is CjBlockExpression) {
                return bodyExpression
            }

            return null
        }
}

